Lib/ufo2ft/__init__.py

```diff
@@ -600,48 +641,155 @@ def compileVariableCFF2(designSpaceDoc, **kwargs):
+    TTFont object.
+
+    .. versionadded:: TODO (Jany) version of addition
+    """
+    kwargs = init_kwargs(kwargs, compileVariableCFF2s_args)
+    excludeVariationTables = kwargs.pop("excludeVariationTables")
     optimizeCFF = CFFOptimization(kwargs.pop("optimizeCFF"))
+    variableFontNames = kwargs.pop("variableFontNames")
 
-    varfont = varLib.build(
-        otfDesignSpace,
+    # Pop inplace because we'll make a copy at this level so deeper functions
+    # don't need to worry
+    inplace = kwargs.pop("inplace")
+    if not inplace:
+        designSpaceDoc = designSpaceDoc.deepcopyExceptFonts()
+
+    vfNameToBaseUfo = _compileNeededSources(
+        kwargs, designSpaceDoc, variableFontNames, compileInterpolatableOTFsFromDS
+    )
+
+    if not vfNameToBaseUfo:
+        logger.warning("No variable fonts to build")
+        return {}
+
+    logger.info(f"Building variable CFF2 fonts: {', '.join(vfNameToBaseUfo)}")
+
+    vfNameToTTFont = varLib.build_many(
+        designSpaceDoc,
         exclude=excludeVariationTables,
         # NOTE optimize=False won't change anything until this PR is merged
         # https://github.com/fonttools/fonttools/pull/1979
         optimize=optimizeCFF >= CFFOptimization.SPECIALIZE,
-    )[0]
-
-    return call_postprocessor(
-        varfont,
-        baseUfo,
-        glyphSet=None,
-        **kwargs,
-        optimizeCFF=optimizeCFF >= CFFOptimization.SUBROUTINIZE,
+        skip_vf=lambda vf_name: variableFontNames and vf_name not in variableFontNames,
     )
+
+    for vfName, varfont in list(vfNameToTTFont.items()):
+        vfNameToTTFont[vfName] = call_postprocessor(
+            varfont,
+            vfNameToBaseUfo[vfName],
+            glyphSet=None,
+            **kwargs,
+            optimizeCFF=optimizeCFF >= CFFOptimization.SUBROUTINIZE,
+        )
+
+    return vfNameToTTFont
+
+
+def _compileNeededSources(
+    kwargs, designSpaceDoc, variableFontNames, compileInterpolatableFunc
+):
+    # We'll need to map <source> elements to TTFonts, to do so make sure that
+    # each <source> has a name.
+    ensure_all_sources_have_names(designSpaceDoc)
+
+    # Go through VFs to build and gather list of needed sources to compile
+    interpolableSubDocs = [
+        subDoc for _location, subDoc in splitInterpolable(designSpaceDoc)
+    ]
+    vfNameToBaseUfo = {}
+    sourcesToCompile = set()
+    for subDoc in interpolableSubDocs:
+        for vfName, vfDoc in splitVariableFonts(subDoc):
+            if variableFontNames is not None and vfName not in variableFontNames:
+                # This VF is not needed so we don't need to compile its sources
+                continue
+            vfNameToBaseUfo[vfName] = vfDoc.findDefault().font
+            for source in vfDoc.sources:
+                sourcesToCompile.add(source.name)
+
+    # Match sources to compile to their Descriptor in the original designspace
+    sourcesByName = {}
```

Comment:
you could populate this `sourcesByName` dict as you go along in the previous nested for loops, the source.name is unique, isn't it? Then you obtain `sourcesToCompile` as set(sourcesByName.keys())`

Reply:
I think the point is to aquire the SourceDescriptors of the original DS?

Reply:
Yes indeed, in the loops above we're iterating the sources of the VF-specific sub-docs, but in sourcesByName we want pointers to the sources of the initial big doc with everything. The sources of the VF-specific sub-docs are different objects (because they have e.g. trimmed down location data etc)